src/client/mod.rs

```diff
@@ -536,3 +559,100 @@ impl WeakClient {
         self.inner.upgrade().map(|inner| Client { inner })
     }
 }
+
+#[cfg(feature = "csfle")]
+pub struct EncryptedClientBuilder<const KMS_SET: bool> {
+    options: ClientOptions,
+    enc_options: AutoEncryptionOptions,
+}
+
+#[cfg(feature = "csfle")]
+impl<const KMS_SET: bool> EncryptedClientBuilder<KMS_SET> {
+    /// Add configuration for a single KMS provider.
+    pub fn kms_provider(
+        mut self,
+        provider: KmsProvider,
+        config: bson::Document,
+    ) -> EncryptedClientBuilder<true> {
+        self.enc_options.kms_providers.insert(provider, config);
+        EncryptedClientBuilder {
+            options: self.options,
+            enc_options: self.enc_options,
+        }
+    }
+
+    /// Add configuration for multiple KMS providers.  Already-added configuration for other
+    /// providers will be preserved.
+    pub fn extend_kms_providers(
+        mut self,
+        providers: impl IntoIterator<Item = (KmsProvider, bson::Document)>,
+    ) -> EncryptedClientBuilder<true> {
+        self.enc_options.kms_providers.extend(providers);
+        EncryptedClientBuilder {
+            options: self.options,
+            enc_options: self.enc_options,
+        }
+    }
+
+    pub fn key_vault_client(mut self, client: impl Into<Option<Client>>) -> Self {
+        self.enc_options.key_vault_client = client.into();
+        self
+    }
+
+    pub fn schema_map(mut self, map: impl IntoIterator<Item = (String, Document)>) -> Self {
+        self.enc_options.schema_map = Some(map.into_iter().collect());
+        self
+    }
+
+    pub fn bypass_auto_encryption(mut self, bypass: impl Into<Option<bool>>) -> Self {
+        self.enc_options.bypass_auto_encryption = bypass.into();
+        self
+    }
+
+    pub fn extra_options(mut self, extra_opts: impl Into<Option<Document>>) -> Self {
+        self.enc_options.extra_options = extra_opts.into();
+        self
+    }
+
+    /// Add TLS options for a single KMS provider.
+    pub fn tls_options(mut self, provider: KmsProvider, options: TlsOptions) -> Self {
```

Comment:
Would we want to combine these TLS options ones with the methods for setting the KMS provider(s)? That way it wouldn't be possible to specify TLS options for a provider that wasn't specified.

If we don't go that route, then I think we'll want to document what happens in that case in the docstring.

Reply:
If we go the route of making `AutoEncryptionOptions` `pub(crate)`, what do you think about unifying them, since the correspondence with the options would matter less in that case?

In general, I think I would be a bit surprised as a user if the bulk and single version of the same method had different functionality, so I'd be in favor of updating the `extend` ones regardless.